clang-tools-extra/clangd/refactor/tweaks/InlineConceptRequirement.cpp

```diff
@@ -0,0 +1,276 @@
+  const auto *ParentDeclaration =
+      ConceptSpecializationExpressionTreeNode->Parent->ASTNode.get<Decl>();
+  if (!isa_and_nonnull<FunctionTemplateDecl>(ParentDeclaration) &&
+      !isa_and_nonnull<FunctionDecl>(ParentDeclaration)) {
+    return false;
+  }
+
+  const FunctionTemplateDecl *FunctionTemplateDeclaration =
+      std::get<0>(findDeclaration<FunctionTemplateDecl>(*Root));
+  if (!FunctionTemplateDeclaration) {
+    return false;
+  }
+
+  auto TemplateArguments =
+      ConceptSpecializationExpression->getTemplateArguments();
+  if (TemplateArguments.size() != 1) {
+    return false;
+  }
+
+  auto TemplateParameterIndex =
+      getTemplateParameterIndexOfTemplateArgument(TemplateArguments[0]);
+  if (!TemplateParameterIndex) {
+    return false;
+  }
+
+  TemplateTypeParameterDeclaration = dyn_cast_or_null<TemplateTypeParmDecl>(
+      FunctionTemplateDeclaration->getTemplateParameters()->getParam(
+          *TemplateParameterIndex));
+  if (!TemplateTypeParameterDeclaration->wasDeclaredWithTypename()) {
+    return false;
+  }
+
+  RequiresToken =
+      findToken(Inputs.AST, FunctionTemplateDeclaration->getSourceRange(),
+                tok::kw_requires);
+  if (!RequiresToken) {
+    return false;
+  }
+
+  return true;
+}
+
+Expected<Tweak::Effect>
+InlineConceptRequirement::apply(const Selection &Inputs) {
+  auto &Context = Inputs.AST->getASTContext();
+  auto &TokenBuffer = Inputs.AST->getTokens();
+
+  tooling::Replacements Replacements{};
+
+  if (auto Err =
+          Replacements.add(generateTemplateParameterReplacement(Context))) {
+    return std::move(Err);
+  }
+
+  auto RequiresReplacement = generateRequiresReplacement(Context);
+
+  if (std::holds_alternative<llvm::Error>(RequiresReplacement)) {
+    return std::move(std::get<llvm::Error>(RequiresReplacement));
+  }
+
+  if (auto Err = Replacements.add(
+          std::get<tooling::Replacement>(RequiresReplacement))) {
+    return std::move(Err);
+  }
+
+  if (auto Err =
+          Replacements.add(generateRequiresTokenReplacement(TokenBuffer))) {
+    return std::move(Err);
+  }
+
+  return Effect::mainFileEdit(Context.getSourceManager(), Replacements);
+}
+
+auto InlineConceptRequirement::getTemplateParameterIndexOfTemplateArgument(
+    const TemplateArgument &TemplateArgument) -> std::optional<int> {
+  if (TemplateArgument.getKind() != TemplateArgument.Type)
+    return {};
+
+  auto TemplateArgumentType = TemplateArgument.getAsType();
+  if (!TemplateArgumentType->isTemplateTypeParmType())
+    return {};
+
+  const auto *TemplateTypeParameterType =
+      TemplateArgumentType->getAs<TemplateTypeParmType>();
+  if (!TemplateTypeParameterType)
+    return {};
+
+  return TemplateTypeParameterType->getIndex();
+}
+
+auto InlineConceptRequirement::generateRequiresReplacement(ASTContext &Context)
+    -> std::variant<tooling::Replacement, llvm::Error> {
+  auto &SourceManager = Context.getSourceManager();
+
+  auto RequiresRng =
```

Comment:
```suggestion
  auto RequiresRange =
```
`Rng` sounds misleading...

Reply:
Renamed to `RequiresRange`